ref: document `dvc artifacts get` (#4809)
iterative/dvc.org: content/docs/api-reference/artifacts_show.md

# dvc.api.artifacts_show()

> Get the path and Git revision for an artifact tracked in a <abbr>DVC

Reviewer: can artifact here be a link to something? (to give ppl context) or an `abbr`

Reviewer 2: I think it might be out of scope of this PR, but we can make a follow up (I can take it) to add a glossary item and add more background to `cmd-ref/artifacts/index` to explain what is an `artifact`.

Reviewer 2: Opened https://github.com/iterative/dvc.org/pull/4844

Reviewer 2: @pmrowla Could you update the PR to use `<abbr>artifact</abbr>` and `<abbr>model registry</abbr>`?

repository</abbr>.

```py
def artifacts_show(
    name: str,
    version: Optional[str] = None,
    stage: Optional[str] = None,
    repo: Optional[str] = None,
) -> Dict[str, str]:
```

## Usage:

```py:
import dvc.api

artifact = dvc.api.artifacts_show(
    'text-classification',
    repo='https://github.com/iterative/example-get-started.git',
)
```

## Description

> Returns a path and Git revision for a named artifact which can then be used in

Reviewer: Also, should be a link here to what those artifacts even are ...

other Python API calls.

The returned dictionary will be of the form:

```py
{
    'path': 'model.pkl',
    'rev': 'c7c6ae0',
}
```

where `path` contains the relative path to the artifact in the DVC repository,
and `rev` contains the Git revision for the specified artifact version or stage.

When neither `version` nor `stage` are provided, the Git revision for the latest
version of the model will be returned.

## Parameters

> - `name` (required) - name of the artifact.

Reviewer 2: We should at least mention how to address artifacts stored outside the root `dvc.yaml`.

- `version` - version of the artifact (mutually exclusive with `stage`).

- `stage` - stage of the artifact (mutually exclusive with `version`).

- `repo` - the location of the DVC project. It can be a URL or a file system
  path. Both HTTP and SSH protocols are supported for online Git repos (e.g.
  `[user@]server:project.git`). _Default_: The current project (found by walking
  up from the current working directory tree).

## Example: Read the contents of an artifact

```py
import pickle
import dvc.api

> artifact = dvc.api.artifacts_show(

Reviewer: @dberenbaum  is there a plan to have a single API that is similar to the `artifact get`? Also need to think on how to pass the Studio token into it (e.g. I'm using Colab to download a model).

Reviewer 2: No, we decided to cut the scope to drop that in https://github.com/iterative/dvc/issues/9100#issuecomment-1663975660. @pmrowla raised the idea to cut the scope and I agree for a few reasons (and I think he aligned with you here, although I know it's changed many times):
1. UX for directories becomes more complicated in the Python API.
2. There are other options to download models. Even assuming it's important to be able to download from Colab, you still could either download via CLI or set storage creds and use the Python API (not sure it's any more work than setting the Studio token). 
3. We don't even have feedback on the most basic model download scenarios yet, and I'd rather wait for that before doing more.

Reviewer: @dberenbaum tbh, it's was my basic assumption that people in 50% cases download a model on CI, and in 50% download it when the server starts. So I was assuming that's is a basic scenario, similar to CLU. We'll be releasing only CLI part for now? and do Python as a followup? Or wait for the Python one? 

> UX for directories becomes more complicated in the Python API.

can it be a simple as - fail for directories + a an API that could list the files? Does the complexity come from DVC (that are are trying to fit into DVC itself), or somewhere else?

Author: The API discussion and decision to hold off on doing anything else for now stems from this comment https://github.com/iterative/dvc/issues/9100#issuecomment-1657596765

It's already possible to list the file paths with
```py
artifact = dvc.api.artifacts_show(
    'text-classification',
    stage='prod',
    repo='https://github.com/iterative/example-get-started.git',
)
fs = dvc.api.DVCFileSystem(
    'https://github.com/iterative/example-get-started.git',
    rev=artifact['rev'],
)
fs.ls(artifact['path'])
```

The main issue I had with a single new call for even something like listing paths for files in the (directory) artifact is that as soon as you want to do anything with those paths you need to use the existing path-based DVC API anyways.

Reviewer: Yep, it might be fine for Studio to also support file-path based APIs, wdyt? (sorry if I'm repeating some of the arguments from the existing discussion).

Author: I guess the question is what do we think users will want to do with the Studio API?

You can already get the download URLs (and a file listing for dirs) with the existing `get-download-uris` call. From there the user can use whatever their python HTTP client of choice is (like `requests`) to download/stream/etc the files that they want. I'm not sure we really need to provide any additional API on top of that.

Author: I see that the docs PR for `get-download-uris` was closed, but IMO it would make sense for that to be a publicly documented API call

Author: > Can we add the signed urls as an additional output of what dvc.api.artifact_show returns? Or show those signed urls in dvc.api.get_url when the studio token is available?

I still don't understand the decision to not document the actual Studio REST API call for this. If we just document the API call, anyone can use it and they aren't restricted to using Python + `dvc.api`

Author: > I guess the question is what do we think users will want to do with the Studio API?

I think we still have to answer this question. What is the purpose of exposing the signed URLs through `dvc.api`?

If we want to provide the URL so that the user can then use the tool/programming language of their to stream/download/etc from the HTTP URL, then there is no point in requiring them to use `dvc.api` or `dvc get-url` in the first place (and we should just document the REST API call).

If we think the user is just going to want to download/stream/etc the artifact via DVC, there is no point in exposing the URL. At that point the studio vs DVC remote distinction is just an implementation detail, and the user can already do what they are looking for with `dvc artifacts get`.

Reviewer 2: > I guess the question is what do we think users will want to do with the Studio API?

👍 Good question. The only scenario I know is if I want to stream the artifact to Python and I have access to the Studio token but not the credentials for the underlying storage. I'm not sure that's enough benefit to expose it yet, but if it just means merging the docs PR, I'm fine with it. I feel it may be a bit distracting/confusing to have so many ways to get models, but it's not a blocker for me.

Maybe @shcheklein has some other user story in mind.

Reviewer 2: Snippet for using the Studio API directly:

```python
import json
import os
import requests

token = os.environ["DVC_STUDIO_TOKEN"]
url = "https://studio.iterative.ai/api/model-registry/get-download-uris"
headers = {"Authorization": f"token {token}"}
params = {
    "repo": "git@github.com:iterative/demo-bank-customer-churn.git",
    "name": "randomforest-model",
    "version": "v2.0.0"
}

response = requests.get(url, headers=headers, params=params)
for relpath, url in json.loads(response.content).items():
   ...
```

Pseudo-snippet with a DVC API:

```python
import dvc.api
import os

token = os.environ["DVC_STUDIO_TOKEN"]
params = {
    "repo": "git@github.com:iterative/demo-bank-customer-churn.git",
    "name": "randomforest-model",
    "version": "v2.0.0",
    "token": token
}

for relpath, url in dvc.api.get_artifact_urls(**params):
    ...
```

Is either of these close to what you have in mind? 

Pros and cons of each to my mind:
1. Using the Studio API is not that much more complex and doesn't even require DVC or any other dependency to be installed.
2. The Studio API requires documenting a separate API which adds another layer and is probably more complex than including it in `dvc.api.artifacts_show` or some other DVC API method.

Author: The original docs PR for the Studio API contained examples using `curl`, which allows using the Studio API directly from CLI (without DVC at all). It could just as easily have additional snippets for Python and whatever other programming languages we want. Like @dberenbaum noted, using the Studio REST API directly is not any more complicated than using `dvc.api`.

Reviewer: > Using the Studio API is not that much more complex and doesn't even require DVC or any other dependency to be installed.

Good point btw. that alone is a very strong reason to use Studio API (or some lightweight wrapper (dvc-studio-client?)).

Can it be something like:

```python
with studio.api.artifact("randomforest-model", token=optional) as model:
   # do something with a model (if it's a directory it downloads a file with links and people would need to use requests) 
```

- it can read token inside
- it doesn't require repo name by default, neither version, revision, etc

wdyt?
- it should not be doing clones, etc (I think that we agree already)
- to your point, it should not be part of `dvc.api` (to heavy to install, too studio specific)

Author: > it doesn't require repo name by default

The Studio API requires the repo URL, so all of the existing dvc/dvc-studio-client calls require it as well. If it's possible for Studio to determine the repo from the token, this needs to be handled on the Studio side (and the URL parameter should be removed from the REST API call).

> neither version, revision, etc

These are already optional in the Studio API. It will return what Studio considers to be the latest model if you don't provide them

If you are looking for a one-liner kind of thing you can already do
```py
from dvc_studio_client.model_registry import get_download_uris

for filename, uri in get_download_uris("https://github.com/model/registry.git", "randomforest-model"):
    ...
```

`dvc_studio_client.model_registry.get_download_uris()` is essentially just a wrapper for the Python `requests` snippet @dberenbaum posted. The only difference is it automatically reads `os.environ[DVC_STUDIO_TOKEN]`. But in this case, the user still has to use `requests` to download/stream the model anyways, so IMO it would be better for us to just document the entire workflow using `requests` and not `dvc_studio_client`.

Reviewer: > The Studio API requires the repo URL, so all of the existing dvc/dvc-studio-client calls require it as well. If it's possible for Studio to determine the repo from the token, this needs to be handled on the Studio side (and the URL parameter should be removed from the REST API call).

folks, to clarify. I'm not looking into this from DVC or Studio perspective. So it might be indeed less relevant to this PR. I was trying to understand the plan here, and scope of is being done here a bit better (e.g. what does that API do internally, can we indeed use it alone, or not, etc).

> if you are looking for a one-liner kind of thing you can already do

right, the difference is that in my case I would make it return the model object itself right away if it's a single file (again, not a change for this PR, but more of a product discussion for the future if we decide to do this)

Anyways, if that direct requests snippet works, if we can make it part of the Studio FE and people will be able to copy paste and run - it's probably fine to start with.

Reviewer 2: That all sounds good. Just to add my two cents and maybe explain why I have been less aggressive on pushing for a particular solution here:

> right, the difference is that in my case I would make it return the model object itself right away if it's a single file

It feels like the ML frameworks and deployment options are so varied as to whether they expect to read a url, an object, etc. that I'm not even sure which is the "happy path."

Get the path and Git revision for an <abbr>artifact</abbr> tracked in a
<abbr>DVC repository</abbr>.

```py
def artifacts_show(
    name: str,
    version: Optional[str] = None,
    stage: Optional[str] = None,
    repo: Optional[str] = None,
) -> Dict[str, str]:
```

## Usage:

```py:
import dvc.api

artifact = dvc.api.artifacts_show(
    'text-classification',
    repo='https://github.com/iterative/example-get-started.git',
)
```

## Description

Returns a path and Git revision for a named artifact which can then be used in
other Python API calls.

The returned dictionary will be of the form:

```py
{
    'path': 'model.pkl',
    'rev': 'c7c6ae0',
}
```

where `path` contains the relative path to the artifact in the DVC repository,
and `rev` contains the Git revision for the specified artifact version or stage.

When neither `version` nor `stage` are provided, the Git revision for the latest
version of the model will be returned.

## Parameters

- `name` (required) - name of the artifact. By default DVC will search for
  artifacts declared in a `dvc.yaml` file located at the root of the DVC
  repository. Artifacts declared in other `dvc.yaml` files should be addressed

>   in the form `path/to/dvc.yaml:artifact_name`.

Reviewer 2: ```suggestion
  in the form `path/to/dvc.yaml:artifact_name` or `path/to:artifact_name` (where `dvc.yaml` is omitted).
```

Not sure we need a whole admon about this

>   in the form `path/to/dvc.yaml:artifact_name`.
> 
> <admon icon="tip">
> 
> DVC also accepts <abbr>model registry</abbr>/[GTO](/doc/gto) style artifact
> names in the form `path/to:artifact_name` (where `dvc.yaml` is omitted).
> 
> </admon>

Reviewer 2: ```suggestion
  in the form `path/to/dvc.yaml:artifact_name` or `path/to:artifact_name` (where `dvc.yaml` is omitted).
```

Don't think it's worth a whole admon

# dvc.api.artifacts_show()

Get the path and Git revision for an

> [artifact](/doc/user-guide/project-structure/dvcyaml-files#artifacts) tracked in

Reviewer 2: ```suggestion
<abbr>artifact</abbr> tracked in
```

a <abbr>DVC repository</abbr>.

```py
def artifacts_show(
    name: str,
    version: Optional[str] = None,
    stage: Optional[str] = None,
    repo: Optional[str] = None,
) -> Dict[str, str]:
```

## Usage:

```py:
import dvc.api

artifact = dvc.api.artifacts_show(
    'text-classification',
    repo='https://github.com/iterative/example-get-started.git',
)
```

## Description

Returns a path and Git revision for a named artifact which can then be used in
other Python API calls.

The returned dictionary will be of the form:

```py
{
    'path': 'model.pkl',
    'rev': 'c7c6ae0',
}
```

where `path` contains the relative path to the artifact in the DVC repository,
and `rev` contains the Git revision for the specified artifact version or stage.

When neither `version` nor `stage` are provided, the Git revision for the latest
version of the model will be returned.

## Parameters

- `name` (required) - name of the artifact. By default DVC will search for
  artifacts declared in a `dvc.yaml` file located at the root of the DVC
  repository. Artifacts declared in other `dvc.yaml` files should be addressed
  in the form `path/to/dvc.yaml:artifact_name`.

<admon icon="tip">

> DVC also accepts GTO/Studio Model Registry style artifact names in the form

Reviewer 2: ```suggestion
DVC also accepts <abbr>model registry</abbr>/[GTO](/doc/gto) style artifact names in the form
```
